### ISSUE TITLE
Tasks using DependsOn with a background task will never start

### ISSUE THREAD

**Issue by pandorla1984** [Microsoft/vscode#70283]
- VSCode Version: 1.32.1
- OS Version: 18309

Steps to Reproduce:

create launch.json:
`{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [        
        {
            "type": "chrome",
            "request": "launch",
            "name": "Launch Chrome against localhost",
            "url": "http://localhost:8080",
            "webRoot": "${workspaceFolder}\\src\\FrontEnd\\build",
            "preLaunchTask": "core"
        }
    ]
}`
2. create task.json:
`{
    // See https://go.microsoft.com/fwlink/?LinkId=733558 
    // for the documentation about the tasks.json format
    "version": "2.0.0",
    "tasks": [
        {
            "label": "core",            
            "type": "npm",
            "script": "start",
            "path": "src/FrontEnd/core/",
            "isBackground": true,
            "dependsOn":["loss"],
            "problemMatcher":{
                "owner": "custom",
                "pattern":[
                    {
                        "regexp": "something not exists",
                        "file": 1,
                        "location": 2,
                        "message": 3
                    }
                ],
                "background": {
                    "activeOnStart": true,
                    "beginsPattern": "core@1.0.0 start",
                    "endsPattern": "created ..\\\\build\\\\vendor.js"
                }
            }
        },
        {
            "label": "loss",            
            "type": "npm",
            "script": "watch",
            "path": "src/FrontEnd/loss/",
            "isBackground": true,            
            "problemMatcher":{
                "owner": "custom",
                "pattern":[
                    {
                        "regexp": "something not exists",
                        "file": 1,
                        "location": 2,
                        "message": 3
                    }
                ],
                "background": {
                    "activeOnStart": true,
                    "beginsPattern": "loss@1.0.0 start",
                    "endsPattern": "waiting for changes"
                }
            }
        }
    ]
}`

I can use beginsPattern/endsPattern with no issue if I direct launched by preLaunchTask, but if I have DependsOn for same task, it will run the dependencies but can't detect it's status. it looks like a bug for me. 

I tried to use the new Terminal: "terminal.integrated.windowsEnableConpty": false/true, no difference. 
I tried to use presentation.panel:"dedicated", no difference. 

I think we have done a nice work on putting tasks running in background and launch, but not those tasks in DependsOn property. 

Does this issue occur when all extensions are disabled?: Yes

**Comment by pandorla1984** [comment]
no, 

> (Experimental duplicate detection)
> Thanks for submitting this issue. Please also check if it is already covered by an existing one, like:
> 
> * [Background tasks are not working properly (#63836)](https://www.github.com/Microsoft/vscode/issues/63836)

no not a duplicate.

**Comment by StephenWeatherford** [comment]
As far as I can tell, there's currently no way to support a task that runs multiple watch tasks at the same time.  I wasted several hours yesterday trying to get this to work.  Am I wrong?

**Comment by dbaeumer** [comment]
@StephenWeatherford what is supported is having one main task that starts n other tasks that run in watch mode. Something like:

```json
		{
			"label": "watch",
			"dependsOn": [
				"watch:types",
				"watch:jsonrpc",
				"watch:protocol",
				"watch:server",
				"watch:client"
			],
			"problemMatcher": []
		}
```

were `watch:*` tasks are all background tasks.

What doesn't work right now is to serialize background tasks. They reason is that is it not clear to define when a background task is in a state that it depending task can start.

**Comment by AlexAegis** [comment]
I also wasted a few hours trying to automate a docker compose to run before running my server. They work independently, but I can't depend on one with another, it just won't advance to the next task even though the problemMatcher/endsPattern matches.

**Comment by alexr00** [comment]
@AlexAegis Are you on Windows and if so, what version? And what version of VS Code? There were some terminal/OS problems that were making it so lines wrapped unexpectedly in the terminal and problem matchers would become unreliable.

**Comment by AlexAegis** [comment]
Yes I'm on Windows 10 version 1903 build no 18362.295, and 1.37.1 (User) of VS Code. I haven't changed any terminal settings, the default plain PowerShell.

**Comment by alexr00** [comment]
Can you share your task and your problem matcher?

**Comment by ProLoser** [comment]
I'm having this problem too:

```
// tasks.json
{
    // See https://go.microsoft.com/fwlink/?LinkId=733558 
    // for the documentation about the tasks.json format
    "version": "2.0.0",
    "tasks": [
        {
            "label": "Jester Web",
            "type": "npm",
            "script": "jester-web",
            "problemMatcher": [],
            "isBackground": true,
            "dependsOn": ["UVP Web"],
            "presentation": {
                "echo": true,
                "reveal": "always",
                "focus": true,
                "panel": "dedicated",
                "showReuseMessage": true,
                "clear": true,
                "group": "Dev Servers"
            }
        },
        {
            "label": "UVP Web",
            "type": "npm",
            "script": "web",
            "isBackground": true,
            "presentation": {
                "echo": true,
                "reveal": "always",
                "focus": true,
                "panel": "dedicated",
                "showReuseMessage": true,
                "clear": true,
                "group": "Dev Servers"
            },
            "problemMatcher": {
                "pattern":{
                    "regexp": "whatever",
                    "file": 1,
                    "location": 2,
                    "message": 3,
                },
                "background": {
                    "activeOnStart": true,
                    "beginsPattern": "Starting 'buildWebappEsm'",
                    "endsPattern": "Starting 'devWatch'"
                }
            }
        },
    }
}
```

And this is the UVP Web task's log output (note that I had to kill the watching task for the right pane to start):

![image](https://user-images.githubusercontent.com/67395/67438454-650a5e80-f5a8-11e9-8f40-902bc9aac7b5.png)

`Starting 'devWatch'` is not being picked up and allowing the `Liteframes` task to start. I had to manually kill the left pane task before the right one would start.

**Comment by danielmahon** [comment]
### My use case

Have tried multiple iterations of configuring the "watch" command, all which run but prevent the debug from continuing. Only works if the watch command is the only task defined in `preLaunchTask`.

launch.json
```js
{
  // Use IntelliSense to learn about possible attributes.
  // Hover to view descriptions of existing attributes.
  // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
  "version": "0.2.0",
  "configurations": [
    {
      "name": "Flutter",
      "request": "launch",
      "type": "dart",
      "flutterMode": "debug",
      "postDebugTask": "cleanup",
      // Doesnt work
      "preLaunchTask": "build",
      // Works! but doesnt allow for multiple tasks
      "preLaunchTask": "flutter: build_runner watch"
    }
  ]
}
```

tasks.json
```js
{
  // See https://go.microsoft.com/fwlink/?LinkId=733558
  // for the documentation about the tasks.json format
  "version": "2.0.0",
  "tasks": [
    {
      "label": "cleanup",
      "command": "echo ${input:terminate}",
      "type": "shell"
    },
    {
      "label": "build:graphql",
      "command": "graphql get-schema -e development",
      "type": "shell"
    },
    // This gives same result
    // {
    //   "label": "build:flutter",
    //   "command": "watch",
    //   "type": "flutter",
    //   "isBackground": true,
    //   "problemMatcher": ["$dart-pub-build_runner"]
    // },
    {
      "label": "build",
      // Both tasks run, but debug doesnt start (tried multiple variations)
      "dependsOn": ["build:graphql", "flutter: build_runner watch"],
      "dependsOrder": "sequence"
    }
  ],
  "inputs": [
    {
      "id": "terminate",
      "type": "command",
      "command": "workbench.action.tasks.terminate",
      "args": "terminateAll"
    }
  ]
}
```

**Comment by LengZai** [comment]
@alexr00 
I got the same issue as the first post with the most recent version of Vscode

- VSCode Version: 1.43.2
- OS Version: Win10 v1909 build 18363.720

Here my launch.json:
```
{
    "version": "0.2.0",
    "configurations": [
      {
        "name": "Chrome",
        "type": "chrome",
        "request": "launch",
        "preLaunchTask": "start-all",
//If i put: "preLaunchTask": "Start httpserver", chrome will launch after launching http-server.
        "url": "http://localhost:4200/#",
        "webRoot": "${workspaceFolder}",
        "sourceMapPathOverrides": {
          "webpack:/*": "${webRoot}/*",
          "/./*": "${webRoot}/*",
          "/src/*": "${webRoot}/*",
          "/*": "*",
          "/./~/*": "${webRoot}/node_modules/*"
        },
        "internalConsoleOptions": "neverOpen"
      },
}
```

and here my tasks.json:
```
{
    "version": "2.0.0",
    "tasks": [
      {
        //"type": "npm",
        //"script": "start-httpserver",
        "type": "shell",
        "command": "node node_modules/http-server/bin/http-server ./dist/PublishDir -p 4000 --cors",
        "isBackground": true,
        "label": "Start httpserver",
        "presentation": {
          "reveal": "always",
          "focus": false,
          "panel": "new",
          "showReuseMessage": true,
          "clear": false,
          "group": "A"
        },
        "problemMatcher": {
          "owner": "http-server",
          "pattern": [
            {
              "regexp": ".",
              "file": 1,
              "location": 2,
              "message": 3
            }
          ],
          "background": {
            "activeOnStart": true,
            "beginsPattern": ".",
            "endsPattern": "."
          }
        }
      },
      {
        "type": "npm",
        "script": "start",
        //npm start = ng serve
        "label": "Start App",
        "isBackground": true,
        "presentation": {
          "focus": true,
          "reveal": "always",
          "panel": "new",
          "group": "A"
        },
        "group": {
          "kind": "build",
          "isDefault": true
        },
        "problemMatcher": {
          "owner": "typescript",
          "source": "ts",
          "applyTo": "closedDocuments",
          "fileLocation": [
            "relative",
            "${cwd}"
          ],
          "pattern": "$tsc",
          "background": {
            "activeOnStart": true,
            "beginsPattern": {
              "regexp": "(.*?)"
            },
            "endsPattern": {
              "regexp": "Compiled |Failed to compile."
            },
          }
        }
      },
      {
        "label": "start-all",
        "dependsOrder": "parallel",
        "dependsOn": [ 
          "Start httpserver", "Start App"
        ]
      }
    ]
  }
```

Here what i got on the terminals:
![image](https://user-images.githubusercontent.com/20622767/77813125-0611be00-70e1-11ea-8ed2-31e7231f801c.png)

So if on preLaunchTask, i set a task, chrome will launch, but if i use "dependsOn", it won't start chrome except if i kill the http-server. Even if on "dependsOn" i only put 1 task (i.e.: Start httpserver), It won't work.

I tried multiple vscode version as i believed it works before
https://stackoverflow.com/questions/44242048/how-to-make-vscode-not-wait-for-finishing-a-prelaunchtask#comment105938297_54017304

- vscode 1.31-1-2
- vscode 1.41.1-25
- vscode 1.42.1-27

I still have the same issue. I spent a bunch of days with multiple syntax, solution, stackoverflow, this github issues list. I don't find any solution so i guess it's really a bug behaviour using "dependsOn".
It is not similar to this issue #92792
because on his example, the 2 tasks are not background tasks but simple ping which will stop after 2 - 12 pings as defined on his package.json

In the meantime, I will try npm-run-all module but it will be great to make it work natively with vscode.

Thanks

**Comment by linusbrolin** [comment]
> @StephenWeatherford what is supported is having one main task that starts n other tasks that run in watch mode. Something like:
> 
> ```json
> 		{
> 			"label": "watch",
> 			"dependsOn": [
> 				"watch:types",
> 				"watch:jsonrpc",
> 				"watch:protocol",
> 				"watch:server",
> 				"watch:client"
> 			],
> 			"problemMatcher": []
> 		}
> ```
> 
> were `watch:*` tasks are all background tasks.
> 
> What doesn't work right now is to serialize background tasks. They reason is that is it not clear to define when a background task is in a state that it depending task can start.

@dbaeumer The vscode docs say that the background tasks need problemMatchers that tracks when they are "done":
https://code.visualstudio.com/Docs/editor/tasks#_compound-tasks
![image](https://user-images.githubusercontent.com/3481784/80492129-e4b42400-8963-11ea-8e3a-1f8d969ca7d3.png)
But I cannot get the "One" task to actually start, if task "Two" or "Three" is a watch task, even if I have a problemMatcher that catches when the watch task has finished recompiling.
It just sits there, waiting, until I manually kill the watch task, then the "One" task will do its thing.

**Comment by StephenWeatherford** [comment]
Thanks, that may work for me, will give it a try.

**Comment by penx** [comment]
I created a basic repo that shows the issue of not being able to create a task that depends on a watch task:

https://github.com/penx/vscode-task-dependson

In this project:
- the launch configuration depends on task 'Two' and works fine
- the task 'One' depends on 'Two' but never executes

As per @linusbrolin's [comment above](https://github.com/microsoft/vscode/issues/70283#issuecomment-620606223), the [documentation says that this should work](https://code.visualstudio.com/Docs/editor/tasks#_compound-tasks):

> Any background/watch tasks used in dependsOn with "dependsOrder": "sequence" must have a problem matcher that tracks when they are "done"

... so please can you either mark this ticket as a bug or else update the documentation to make it clear this is not supported?

**Comment by cha0s** [comment]
Just chipping in here that I also found this problem and it would be super amazing if it was made to work.

**Comment by philliphoff** [comment]
We appear to be running into this as well in [vscode-dapr](https://github.com/microsoft/vscode-dapr/issues/97) when layering use of Dapr tasks on top of Azure Functions tasks, both of which are background tasks with begin/end patterns.  I'm not able to make the Dapr task wait for the Azure Function task (as a `dependsOn`) as VS Code doesn't seem to recognize when the Azure Functions task completes.

**Comment by alexr00** [comment]
@penx thanks for the simple repro repository. 

Until now, you could have background tasks in `dependsOn`, but since those background tasks never "finish", you couldn't sequence anything off of them. 

I have pushed a small change that _does_ allow some sequencing. You can have background tasks in `dependsOn` _and_ they can be sequenced. **However** any errors that those background tasks produce will not prevent subsequent tasks from executing. I will leave this issue open to track supporting that.

**Comment by mohsen1** [comment]
@alexr00 how can we specify a background task in a good state for its dependent tasks to kick in? For me, I want to start a sync task that will take some time to sync my files and then outputs `"sync is completed, now watching for file changes..."`. I want to start my dependent tasks to start when that output is printed.

**Comment by tetchel** [comment]
I would expect that the task runner would detect the `endsPattern` of the first task 
https://code.visualstudio.com/docs/editor/tasks#_background-watching-tasks

**Comment by ProLoser** [comment]
@mohsen1 @tetchel in my mind, I planned on having an `endsPattern` detect when a service is finished running through it's initial setup. Like when the watcher is running, or files are being served and are ready to be accessed, that would be matched for my `endsPattern`. I don't really care about exits on those tasks unless I'm killing my services.

**Comment by cha0s** [comment]
It is an interesting question. With a background task, `beginsPattern` and `endsPattern` are somewhat overloaded with the concept of something like "begin background task" and "end background task" (e.g. "files changed", "compilation complete", respectively).

However it probably makes sense to differentiate between "task start" and "task end" and "service start" and "service end" in this context. Perhaps a better solution would be to add two new keys to the `background` task configuration to reflect these two different contexts.

**Comment by ProLoser** [comment]
I don't think all that additional complexity is going to add anything if we
can do what we want with the options we have now.

**Comment by panayot-cankov** [comment]
This:
```
{
    "version": "2.0.0",
    "tasks": [
        {
            "label": "npm watch",
            "type": "npm",
            "script": "watch",
            "problemMatcher": "$tsc-watch",
            "isBackground": true,
            "presentation": {
                "reveal": "always",
                "panel": "dedicated"
            },
            "group": "build"
        },
        {
            "label": "composite watch",
            "dependsOn": [
                "npm watch"
            ]
        }
    ]
}
```
I have a launch with preLaunchTask: "composite watch". Launching with debug will start the typescript compiler, wait for it to compile and then launch the index.js. The typescript watch task continues to execute. If I try debug a second time, the debugger stalls waiting.

The same setup works as expected with preLaunchTask: "npm watch". First time the watcher starts and it takes time to compile, second F5 though is much faster as the compilation is done already.

Ultimately I need to have 3 watchers running in parallel, and F% start a debugger when they are all settled.

**Comment by Luxcium** [comment]
I would like to add my use-case and see if anyone can help:

I have 3 apps stating using « `npm start` » (` "start": "webpack serve"`)

2 of them can be started in parallel `cart` and `products` but the last one `federal-container` must be started only after the first two are running... 

also, I could theoretically run only `cart` and `products` in isolation therefore I need to define tasks for those two anyway... 

is there a way to either `"dependsOrder": "sequence"` on the 3 or I already can `"dependsOrder": "sequence"` and would need the 3rd to sequentially start after the first group...

all are in watch mode or are not completed when the next task should start... 

it all ends with the `compiled successfully` keywords... 

![Screenshot_000001_20221023_145728](https://user-images.githubusercontent.com/42672814/197410486-1a08ecd3-e3c0-4e79-ab69-9cb074eae3fd.png)
